proxy_http.go

```diff
@@ -174,6 +181,11 @@ func (proxy *proxy) processRequests(ctx filters.Context, remoteAddr string, req
 			}
 		}
 
+		if err != nil {
+			// We encountered an error on round-tripping, stop now
+			return err
+		}
```

Comment:
I think the race condition must have been happening when our round-trip upstream times out. When that happens, the Transport is still trying to read from downstream to copy to upstream. But our read loop has already proceeded to the next call to `readRequest`.

The solution is to immediately return the err and not attempt to read any more.